ConfigSpace/c_util.pyx

```diff
@@ -266,83 +204,64 @@ cpdef np.ndarray change_hp_value(
     """
     cdef Hyperparameter current
     cdef str current_name
-    cdef list disabled
-    cdef set visited
-    cdef dict activated_values
     cdef int active
+    cdef int update
     cdef ConditionComponent condition
     cdef int current_idx
     cdef DTYPE_t current_value
+    cdef DTYPE_t target_value
     cdef DTYPE_t default_value
-    cdef list children
-    cdef list children_
-    cdef Hyperparameter ch
-    cdef str child
-    cdef set to_disable
     cdef DTYPE_t NaN = np.NaN
     cdef dict children_of = configuration_space._children_of
 
-    configuration_array[index] = hp_value
-
-    # Hyperparameters which are going to be set to inactive
-    disabled = []
+    # We maintain to_visit as a minimum heap of indices of hyperparameters that may need to be updated.
+    # We assume that the hyperparameters are sorted topologically by their index.
+    to_visit = [index]
+    
+    # Since one hyperparameter may be reachable in several ways, we need to make sure we don't process it twice.
+    scheduled = np.zeros(len(configuration_space), dtype=bool)
+    scheduled[index] = True
 
     # Activate hyperparameters if their parent node got activated
-    children = children_of[hp_name]
-    if len(children) > 0:
-        to_visit = deque()  # type: deque
-        to_visit.extendleft(children)
-        visited = set()  # type: Set[str]
-        activated_values = dict()  # type: Dict[str, Union[int, float, str]]
-
-        while len(to_visit) > 0:
-            current = to_visit.pop()
-            current_name = current.name
-            if current_name in visited:
-                continue
-            visited.add(current_name)
-            if current_name in disabled:
-                continue
-
-            current_idx = configuration_space._hyperparameter_idx[current_name]
+    while len(to_visit) > 0:
+        current_idx = heapq.heappop(to_visit)
+        current_name = configuration_space._idx_to_hyperparameter[current_idx]
+        conditions = configuration_space._parent_conditions_of[current_name]
+
+        active = True
+        for condition in conditions:
+            if not condition._evaluate_vector(configuration_array):
+                active = False
+                break
+
+        update = False
+        if current_idx == index:
+            if not active:
+                raise ValueError(
+                    "Attempting to change the value of the inactive hyperparameter '%s' to '%s'." % (hp_name, hp_value))
```

Comment:
As a nice side effect of the code reorganization, we explicitly prevent the caller from changing the value of an inactive HP.